### PR TITLE
Unused record fields rule

### PR DESCRIPTION
**Pull request by maco**

### CODE REVIEW

File: lib/meandro/meandro_rule.ex [hunk 1]
```diff
@@ -18,7 +18,8 @@ defmodule Meandro.Rule do
           line: non_neg_integer(),
           text: charlist(),
           rule: t(),
-          pattern: ignore_pattern()
+          pattern: ignore_pattern(),
+          module_name: any()
```
Comment (elbrujohalcon) [inline lib/meandro/meandro_rule.ex:22]: Shouldn't it be `atom()` instead of `any()`?

**Review by elbrujohalcon — COMMENTED** [review]
Does it work if somebody uses `Record.defrecord` (i.e. the full name of `defrecord` or `defrecordp`)?

**Comment by maco** [comment]
Right now it crashes ;) It says “WIP” for a reason.

And yes, it should be atom, but I didn’t figure that out until about the
time I posted in Slack that I finally understood the error message that
accompanied the crash, and I haven’t changed the code since.
-- 
Mackenzie Morgan
(she/her)

File: lib/meandro/rules/unused_record_fields.ex [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+      result
+    end
+  end
+
+  @impl Meandro.Rule
+  def is_ignored?(_pattern, _ignore_spec) do
+    false
+  end
+
+  defp analyze_file(file, ast) do
```
Comment (tak30) [inline lib/meandro/rules/unused_record_fields.ex:26]: Should it be named analyze_module?

Comment (maco) [reply lib/meandro/rules/unused_record_fields.ex:26]: probably a rename we need to do in all the rules now that the `{module, file, ast}` split happened.

File: lib/meandro/rules/unused_callbacks.ex [hunk 3]
```diff
@@ -34,33 +36,35 @@ defmodule Meandro.Rule.UnusedCallbacks do
...
     %{callbacks: callbacks} = acc
 
-    for {module, name, arity, line, count} <- callbacks, count == 0 do
+    for {module, name, arity, line, count} <- Enum.reverse(callbacks), count == 0 do
       %Meandro.Rule{
         file: file,
         line: line,
         text: "Callback #{module}:#{name}/#{arity} is not used anywhere in the module",
-        pattern: {name, arity}
+        pattern: {module, name, arity}
```
Comment (elbrujohalcon) [inline lib/meandro/rules/unused_callbacks.ex:47]: Do we need to add `module` to all our ignore patterns? I would expect that to be taken care of externally. That is… I would expect whatever we implement in #33 to first filter by module on its own and then only call `is_ignored?/2` with patterns and keys that belong to the same module.

File: lib/meandro/rules/unused_record_fields.ex [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    for {module, name, _num_of_fields, unused_fields, line} <-
+          Enum.reverse(acc[:records]),
+        unused_fields != [] do
+      camel_name = name |> Atom.to_string() |> Macro.camelize()
+
+      %Meandro.Rule{
+        file: file,
+        line: line,
+        text:
+          "Record :#{name} (#{camel_name}) has the following unused fields in the module: #{inspect(unused_fields)}",
```
Comment (elbrujohalcon) [inline lib/meandro/rules/unused_record_fields.ex:46]: I think it's better to emit one warning per field so they can be ignored *individually*.
The pattern will then be `{name, unused_field}` and people will be able to easily ignore the warning for just one field in the record.

Comment (elbrujohalcon) [reply lib/meandro/rules/unused_record_fields.ex:46]: Or, alternatively, they'll be able to specify just the record name as the ignore pattern and ignore them all if they want to do so.

**Review by elbrujohalcon — CHANGES_REQUESTED** [review]

File: lib/meandro/rules/unused_callbacks.ex (hunk 3 above)
Comment (pablocostass) [reply lib/meandro/rules/unused_callbacks.ex:47]: Yeah that makes sense :)

File: lib/meandro/rules/unused_record_fields.ex (hunk 4 above)
Comment (pablocostass) [reply lib/meandro/rules/unused_record_fields.ex:46]: I like the former suggestion more, albeit I feel like that would clutter the output a bit; all being said though, the `unused_struct_fields` does what you are suggesting, and consistency is the key :)

Comment (elbrujohalcon) [reply lib/meandro/rules/unused_record_fields.ex:46]: And you usually shouldn't (**shouldn't**) have too many unused fields in the same record.

**Review by tak30 — CHANGES_REQUESTED** [review]
We need to look for record usage in other files

**Review by tak30 — APPROVED** [review]

**Review by elbrujohalcon — APPROVED** [review]

**Pull request merged**